core/src/main/java/org/elasticsearch/common/path/PathTrie.java

```diff
@@ -200,18 +300,47 @@ public void insert(String path, T value) {
             return;
         }
         int index = 0;
-        // supports initial delimiter.
+        // Supports initial delimiter.
         if (strings.length > 0 && strings[0].isEmpty()) {
             index = 1;
         }
         root.insert(strings, index, value);
     }
 
+    /**
+     * Insert a value for the given path. If the path already exists, replace the value with:
+     * <pre>
+     * value = updater.apply(oldValue, newValue);
+     * </pre>
+     * allowing the value to be updated if desired.
+     */
+    public void insertOrUpdate(String path, T value, BiFunction<T, T, T> updater) {
```

Comment:
I think that would be worth trying, but it would be a large undertaking. Since this is blocking a breaking change for 6.0, would you be okay with me looking at that in a follow-up?